crates/sui-transactional-test-runner/src/test_adapter.rs

```diff
@@ -584,11 +588,103 @@ impl<'a> MoveTestAdapter<'a> for SuiTestAdapter<'a> {
             }
+            SuiSubcommand::UpgradePackage(UpgradePackageCommand {
+                package,
+                upgrade_capability,
+                gas_budget,
+            }) => {
+                let data = data.ok_or_else(|| {
+                    anyhow::anyhow!(
+                        "Expected a module text block following 'upgrade' starting on lines {}-{}",
+                        start_line,
+                        command_lines_stop
+                    )
+                })?;
+                let (units, warnings_opt) = self.compile_source_units(data.path())?;
+                let modules = units
+                    .into_iter()
+                    .map(|unit| match unit {
+                        AnnotatedCompiledUnit::Module(annot_module) => {
+                            let (named_addr_opt, _id) = annot_module.module_id();
+                            let named_addr_opt = named_addr_opt.map(|n| n.value);
+                            let module = annot_module.named_module.module;
+                            (named_addr_opt, module)
+                        }
+                        AnnotatedCompiledUnit::Script(_) => panic!(
+                            "Expected a module text block, not a script, \
+                                    following 'publish' starting on lines {}-{}",
+                            start_line, command_lines_stop
+                        ),
+                    })
+                    .collect();
+
+                let output =
+                    self.upgrade_package(package, modules, upgrade_capability, gas_budget)?;
+                Ok(merge_output(warnings_opt, output))
+            }
+        }
+    }
+}
+
+fn merge_output(left: Option<String>, right: Option<String>) -> Option<String> {
+    match (left, right) {
+        (None, right) => right,
+        (left, None) => left,
+        (Some(mut left), Some(right)) => {
+            left.push_str(&right);
+            Some(left)
         }
     }
 }
 
 impl<'a> SuiTestAdapter<'a> {
+    fn upgrade_package(
+        &mut self,
+        current_package_object_id: Option<String>,
+        modules: Vec<(Option<Symbol>, CompiledModule)>,
+        upgrade_capability: Option<u64>,
+        gas_budget: Option<u64>,
+    ) -> anyhow::Result<Option<String>> {
+        let current_package_object_id = current_package_object_id
+            .map_or(ObjectID::from_address(AccountAddress::ZERO), |id| {
+                ObjectID::from_str(&id).unwrap()
+            });
+        let upgrade_capability = upgrade_capability.unwrap_or_default();
+        let all_bytes = modules
+            .iter()
+            .map(|(_, module)| {
+                let mut module_bytes = vec![];
+                module.serialize(&mut module_bytes)?;
+                Ok(module_bytes)
+            })
+            .collect::<anyhow::Result<_>>()?;
+        let gas_budget = gas_budget.unwrap_or(DEFAULT_GAS_BUDGET);
+        let dependencies = system_package_ids();
```

Comment:
Note: this only adds system package ids as dependencies (we can add a `--dependencies` argument later if we want full control).